src/nvim/tui/tui.c

```diff
@@ -527,7 +529,13 @@ static void tui_flush(UI *ui)
   }
 
   unibi_goto(ui, data->row, data->col);
+
+  if (!data->busy) {
+    unibi_out(ui, unibi_cursor_normal);
+  }
+
   flush_buf(ui);
+  unibi_out(ui, unibi_cursor_invisible);
```

Comment:
Perhaps I have misunderstood how this works, but can you explain, why the following code would not be enough/work?
My explanation that it would work would be:
Looks like `tui_flush()` is the only place where the cursor is "switched" on (besides tui_stop()).
If we need to show the cursor, we do so and flush. After that the cursor
will be hidden at once, but this change will only be visible with the next call to `flush_buf()`. 
```c
if (!data->busy) {
      unibi_out(ui, unibi_cursor_normal); // Shows cursor
}
flush_buf(ui);
if (!data->busy) {
      unibi_out(ui, unibi_cursor_invisible); // Hides cursor
}
```
So it seems we only need to hide the cursor when the cursor was shown here.

The only problem that I see would be that the buffer is filled (and flushed with `flush_buf()`) before the next call to `tui_flush()` and so the cursor would be hidden until the next `tui_flush()`. But this is also how the old code works.